redfish-core/lib/account_service.hpp

```diff
@@ -130,6 +132,190 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
+    bool isFoundAccountTypes = true;
+
+    bool isRedfish = false;
+    bool isIPMI = false;
+    bool isHostConsole = false;
+    bool isManagerConsole = false;
+    bool isWebUI = false;
+
+    for (const auto& accountType : *accountTypes)
+    {
+        if (accountType == "Redfish")
+        {
+            isRedfish = true;
+        }
+        else if (accountType == "IPMI")
+        {
+            isIPMI = true;
+        }
+        else if (accountType == "WebUI")
+        {
+            isWebUI = true;
+        }
+        else if ((accountType == "HostConsole"))
+        {
+            isHostConsole = true;
+        }
+        else if (accountType == "ManagerConsole")
+        {
+            isManagerConsole = true;
+        }
+        else
+        {
+            // set false if accountTypes not found and return
+            isFoundAccountTypes = false;
+            return isFoundAccountTypes;
+        }
+    }
+
+    if (isRedfish)
+    {
+        userGroup.emplace_back("redfish");
+    }
+    if (isIPMI)
+    {
+        userGroup.emplace_back("ipmi");
+    }
+    if (isWebUI)
+    {
+        userGroup.emplace_back("web");
+    }
+    if ((isHostConsole) ^ (isManagerConsole))
+    {
+        BMCWEB_LOG_ERROR << "HostConsole or ManagerConsole, one of value is "
+                            "missing to set SSH property";
+        isFoundAccountTypes = false;
+        return isFoundAccountTypes;
+    }
+    if ((isHostConsole) && (isManagerConsole))
+    {
+        userGroup.emplace_back("ssh");
+    }
+
+    return isFoundAccountTypes;
+}
+
+inline void translateUserGroup(const std::vector<std::string>* userGroups,
+                               crow::Response& res)
+{
+    if (userGroups == nullptr)
+    {
+        BMCWEB_LOG_ERROR << "userGroups wasn't a string vector";
+        messages::internalError(res);
+        return;
+    }
+    nlohmann::json& accountTypes = res.jsonValue["AccountTypes"];
+    accountTypes = nlohmann::json::array();
+    for (const auto& userGroup : *userGroups)
+    {
+        if (!getAccountTypeFromUserGroup(userGroup, accountTypes))
+        {
+            BMCWEB_LOG_ERROR << "mapped value not for this userGroup value : "
+                             << userGroup;
+            messages::internalError(res);
+            return;
+        }
+    }
+}
+
+inline void translateAccountType(
+    const std::optional<std::vector<std::string>>& accountType,
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const std::string& dbusObjectPath, bool isUserItself, bool isAdminUser)
+{
+    // user can not disable their own Redfish Property.
+    if (isUserItself)
+    {
+        if (auto it = std::find(accountType->cbegin(), accountType->cend(),
+                                "Redfish");
+            it == accountType->cend())
+        {
+            BMCWEB_LOG_ERROR
+                << "user can not disable their own Redfish Property";
+            messages::accountNotModified(asyncResp->res);
```

Comment:
This should send the StrictAccountTypes message.  (See comment above for details.)
We'll have to document that: IBM systems have restrictions on AccountTypes:
1. A user account AccountTypes is required to contain the "Redfish" value.
2. For accounts which have the Administrator role, the AccountTypes is required to contain the "ManagerConsole" and "HostConsole" values.

Reply:
Agree @joseph-reynolds  what do you say?

Reply:
I disagree.  The AccountTypes property is writable, so the PropertyNotWritable message is not appropriate.

The StrictAccountTypes message is intended to inform the caller there was a problem with the AccountTypes they specified.  This will drive them to the docs to find out what is wrong and how to fix it.  Our docs will then tell them admins must have the "ManagerConsole" and "HostConsole" values.  <--- That is what we discussed at the Redfish forum meeting when we added StrictAccountTypes to the spec.

If the AccountTypes is incorrect, the request must fail.

I will ask our information development (docs) team to add this to our BMC account management section.

Reply:
The StrictAccountTypes message is intended when StrictAccountTypes = True and a value that would be enabled because it is tied to another value being set is left out of the request.
 (e.g. HostConsole is passed and ManagerConsole isn't )
 This code is for disabling Redfish, this doesn't have anything to do with StrictAccountTypes = True (e.g. if we added write support to StrictAccountTypes and set it to false, this check and error message would still be there) 
 The StrictAccountTypes  message doesn't make sense here